pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
+	ctx context.Context
 
-	for _, addr := range addrs {
-		if addr == nil {
-			continue
-		}
+	opts Options
 
-		hash, err := mh.Sum((addr).Bytes(), mh.MURMUR3, -1)
-		if err != nil {
-			return nil, nil, err
-		}
-		keys[i] = parentKey.ChildString(base32.RawStdEncoding.EncodeToString(hash))
-		clean[i] = addr
-		i++
-	}
+	cache       cache
+	ds          ds.TxnDatastore
+	subsManager *pstoremem.AddrSubManager
 
-	return keys[:i], clean[:i], nil
-}
+	flushJobCh chan *addrsRecord
+	cancelFn   func()
+	closedCh   chan struct{}
 
-// AddAddr will add a new address if it's not already in the AddrBook.
-func (mgr *dsAddrBook) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
-	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+	gcCurrWindowEnd    int64
+	gcLookaheadRunning int32
 }
 
-// AddAddrs will add many new addresses if they're not already in the AddrBook.
-func (mgr *dsAddrBook) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
-	if ttl <= 0 {
-		return
+var _ pstore.AddrBook = (*dsAddrBook)(nil)
+
+// NewAddrBook initializes a new address book given a Datastore instance, a context for managing the TTL manager,
+// and the interval at which the TTL manager should sweep the Datastore.
+func NewAddrBook(ctx context.Context, store ds.TxnDatastore, opts Options) (ab *dsAddrBook, err error) {
+	var cache cache = new(noopCache)
+	if opts.CacheSize > 0 {
+		evictCallback := func(key interface{}, value interface{}) {
+			value.(*addrsRecord).Reset()
+			arPool.Put(value)
+		}
+		if cache, err = lru.NewARCWithEvict(int(opts.CacheSize), evictCallback); err != nil {
+			return nil, err
+		}
 	}
-	mgr.setAddrs(p, addrs, ttl, ttlExtend)
-}
 
-// SetAddr will add or update the TTL of an address in the AddrBook.
-func (mgr *dsAddrBook) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
-	addrs := []ma.Multiaddr{addr}
-	mgr.SetAddrs(p, addrs, ttl)
+	ctx, cancelFn := context.WithCancel(ctx)
+	mgr := &dsAddrBook{
+		ctx:         ctx,
+		cancelFn:    cancelFn,
+		opts:        opts,
+		cache:       cache,
+		ds:          store,
+		subsManager: pstoremem.NewAddrSubManager(),
+		flushJobCh:  make(chan *addrsRecord, 32),
+		closedCh:    make(chan struct{}),
+	}
+
+	// kick off periodic GC.
+	go mgr.background()
+
+	return mgr, nil
 }
 
-// SetAddrs will add or update the TTLs of addresses in the AddrBook.
-func (mgr *dsAddrBook) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
-	if ttl <= 0 {
-		mgr.deleteAddrs(p, addrs)
-		return
-	}
-	mgr.setAddrs(p, addrs, ttl, ttlOverride)
+func (ab *dsAddrBook) Close() {
+	ab.cancelFn()
+	<-ab.closedCh
 }
 
-func (mgr *dsAddrBook) deleteAddrs(p peer.ID, addrs []ma.Multiaddr) error {
-	// Keys and cleaned up addresses.
-	keys, addrs, err := keysAndAddrs(p, addrs)
-	if err != nil {
-		return err
+func (ab *dsAddrBook) asyncFlush(pr *addrsRecord) {
+	select {
+	case ab.flushJobCh <- pr:
+	default:
+		log.Warningf("flush queue is full; could not flush peer %v", pr.Id.ID.Pretty())
```

Comment:
I'd be surprised if we don't hit this multiple times a second when querying the DHT.

Reply:
This pathway is only used on read, and the chances are slim. GC is running in the background, so this will only occur if the record becomes outdated between the previous purge cycle and the next, which run every 5 mins by default.

Reply:
Yes, I think the write methods should add the entry in the lookahead window if the new earliest expiry happens to be inside the window.